app/cli/startstream.cpp

```diff
@@ -0,0 +1,190 @@
+#include "startstream.h"
+#include "backend/computermanager.h"
+#include "streaming/session.hpp"
+
+#include <QTimer>
+
+#define COMPUTER_SEEK_TIMEOUT 10000
+#define APP_SEEK_TIMEOUT 10000
+
+namespace CliStartStream
+{
+
+enum State {
+    StateInit,
+    StateSeekComputer,
+    StateSeekApp,
+    StateStartSession,
+    StateFailure,
+};
+
+class Event
+{
+public:
+    enum Type {
+        ComputerUpdated,
+        Executed,
+        Timedout,
+    };
+
+    Event(Type type)
+        : type(type), computerManager(nullptr), computer(nullptr) {}
+
+    Type type;
+    ComputerManager *computerManager;
+    NvComputer *computer;
+};
+
+class LauncherPrivate
+{
+    Q_DECLARE_PUBLIC(Launcher)
+
+public:
+    LauncherPrivate(Launcher *q) : q_ptr(q) {}
+
+    void handleEvent(Event event)
+    {
+        Q_Q(Launcher);
+        Session* session;
+        NvApp app;
+
+        switch (event.type) {
+        case Event::Executed:
+            if (m_State == StateInit) {
+                m_State = StateSeekComputer;
+                m_TimeoutTimer->start(COMPUTER_SEEK_TIMEOUT);
+                m_ComputerManager = event.computerManager;
+                q->connect(m_ComputerManager, &ComputerManager::computerStateChanged,
+                           q, &Launcher::onComputerUpdated);
+                // Seek desired computer by both connecting to it directly (this may fail
+                // if m_ComputerName is UUID, or the name that doesn't resolve to an IP
+                // address) and by polling it using mDNS, hopefully one of these methods
+                // would find the host
+                m_ComputerManager->addNewHost(m_ComputerName, false);
+                m_ComputerManager->startPolling();
+                emit q->searchingComputer();
+            }
+            break;
+        case Event::ComputerUpdated:
+            if (m_State == StateSeekComputer) {
+                if (matchComputer(event.computer) && isOnline(event.computer)) {
+                    if (isPaired(event.computer)) {
+                        m_State = StateSeekApp;
+                        m_TimeoutTimer->start(APP_SEEK_TIMEOUT);
+                        m_Computer = event.computer;
+                        m_ComputerManager->stopPollingAsync();
+                        emit q->searchingApp();
+                    } else {
+                        m_State = StateFailure;
+                        emit q->failed(QString("Computer %1 has not been paired").arg(m_ComputerName));
+                    }
+                }
+            }
+            if (m_State == StateSeekApp) {
+                int index = getAppIndex();
+                if (-1 != index) {
+                    m_State = StateStartSession;
+                    m_TimeoutTimer->stop();
+                    app = m_Computer->appList[index];
+                    session = new Session(m_Computer, app, m_Preferences);
```

Comment:
We should do something in the case that the user is already streaming a different game (`m_Computer->currentGameId != 0 && m_Computer->currentGameId != app->id`). Otherwise, we'll hit the Session::exec() assert:
```
    // The UI should have ensured the old game was already quit
    // if we decide to stream a different game.
    Q_ASSERT(m_Computer->currentGameId == 0 ||
             m_Computer->currentGameId == m_App.id);
```

I would suggest an error message like `QString("%1 is already running. Please quit %1 to stream %2.").arg( runningApp.name).arg(app.name)`

Reply:
👍